### DISCUSSION TITLE
What's the reason to discourage the use of query params in the request handler URL?

### DISCUSSION THREAD

**Discussion by pedrobrost** [mswjs/msw#1486] [Q&A] [answered]
I'm trying to test a paginated list so I want to mock the response of two pages like this:

- `/list?page=1`
- `/list?page=2`

I know I can do it adding a condition in the request handler body but I don't see why it would be a bad practice to just make two separate handlers for each query param value. That way I would avoid adding logic inside my test which could potentially be wrong.

I know that for example `nock` allows you to specify the query string like I mentioned above:

<img width="721" alt="image" src="https://user-images.githubusercontent.com/19985470/205086532-b20949c3-4c42-470b-a910-e83a8dca52b9.png">

Of course I'm not saying the other library is better, I really like `msw` but I'm just curious why it's discouraged to do it that way.

**Comment by mattcosta7** [comment, marked as answer]
I could be incorrect, but I believe it stems from express (and express like path mappings)

https://expressjs.com/en/guide/routing.html#route-paths

![Screen Shot 2022-12-01 at 10 08 56 AM](https://user-images.githubusercontent.com/8616962/205088087-f0d6fbce-9730-4639-babe-d70b0afba767.png)

**Reply by mattcosta7** [reply]
https://github.com/mswjs/msw/blob/ab8b1accf5db10b30647759dde8c1585b8dc4c14/src/utils/matching/matchRequestUrl.ts#L1

Since msw uses this same `path-to-regexp` matching library under the hood, and `?` is treated as a regex subset, query params would be strange to use here

**Reply by pedrobrost** [reply]
oh yes that makes sense! It doesn't look like a good idea to make this part of the route path but have you thought about adding this as a separate feature? like adding the possibility to match by query params without having to add logic in the handler?

**Reply by mattcosta7** [reply]
@pedrobrost Do you know of many libraries that do similar?  Most common server frameworks I'm aware of in the js ecosystem treat routes independent from query parameters (as do many client side routers - although not all). 

What value does that give over declaring the query parameters internally to the handler (which should help to ensure that more cases are handled, like a param that doesn't match, or doesn't exist?)

**Reply by pedrobrost** [reply]
yeah I think you are right, I was thinking of this library as something to help write tests but it's not necessarily attached to only that. While testing I would like to have this feature I'm talking about but maybe it shouldn't be a responsibility of this library to solve that. So thanks!

**Reply by kettanaito** [reply]
This is precisely the answer. 

Query parameters are not part of the resource path on the server. They are more similar to the request body than to the request path if anything. Including query parameters in the request, path is a bad idea for request matching because that matching often boils down to strings comparison. Query parameters are not sensitive to order but strings are. 

> like adding the possibility to match by query params without having to add logic in the handler?

This is what I like about MSW: you can build that feature yourself using our API! 

```js
function withQueryParams(params, resolver) {
  return (req, res, ctx) => {
    // Implement the query params matching as you see fit.
    const matches = Object.entries(params).every(([key, value]) => {
      return req.params[key] === value
    })

    if (!matches) {
      return req.passthrough()
    }

    return resolver(req, res, ctx)
  })
}
```

```js
// handlers.js
import { withQueryParams } from './utils'

export const handlers = [
  rest.get('/user', withQueryParams({ id: 'abc-123' }, (req, res, ctx) => {
    return res(ctx.text('hello'))
  }))
]
```

**Reply by donferi** [reply]
Anyone out there has done this `withQueryParams` TypeScript friendly? Been trying for a bit but having difficulties typing the `ResponseResolver`.
